### ISSUE TITLE
coerce doesn't apply to dot-notation options

### ISSUE THREAD

**Issue by ljharb** [yargs/yargs#1021]
Despite #599 being closed, I'm finding that a key with a dot in it is not getting the coercion function applied (just like demandOption is not working in #995).

**Comment by ljharb** [comment]
This seems to be a bug in yargs-parser; specifically, the `applyCoercions` function doesn't recurse into objects and apply to dot-notation keys?

**Comment by bcoe** [comment]
@ljharb taking a look at this issue, this was a design decision (and different than the issue described in #995).

To apply coercion to an object, it's applied to the top level key:

```js
    it('applies coercion function to all dot options', function () {
      var parsed = parser(['--foo.bar', 'nananana'], {
        coerce: {
          foo: function (val) {
            val.bar += ', batman!'
            return val
          }
        }
      })
      parsed.foo.bar.should.equal('nananana, batman!')
    })
```

tldr; I see dot notation as a property of the top level key:

`foo.bar` indicates that the key `foo` is an object, which happens to have the property `bar`. You can then apply coercions to `foo` and modify the object accordingly.

I think I'll likely revisit #995 with a similar eye; I don't like the idea of being able to configure options for `foo.x`, `foo.y`, because you end end up adding a lot of complexity to the parser (should I be able to configure `foo.x.y`? I'd rather avoid this question all together).

**Comment by ljharb** [comment]
@bcoe I tried that; `.coerce('foo', fn)` makes `foo` show up in the help output, so I don't think that's a viable solution.

**Comment by ljharb** [comment]
I think that if dot notation options are allowed, they should work with everything else; if they aren't going to, then why have that option in the first place?

Is there any way I can configure my `yargs` instance to apply coercions to a nested property, without altering the help output/list of available commands?

**Comment by bcoe** [comment]
@ljharb I'd like to put a stake in the ground and say that we'll never allow specific keys on an object to be configured, e.g., "foo.bar"; this adds a lot of complexity, and I believe _a lot_ of yargs' functionality will not work as expected -- dot notation should be thought of as a way of saying, the argument "foo" is an object, with these keys on it "a", "b", "c", etc. Rather than a way of saying "foo.a" is a distinct option unto itself.

>  I tried that; .coerce('foo', fn) makes foo show up in the help output.

Out of curiosity, why don't you want the option `foo` to show up in help output? I'd definitely entertain the idea of making it so that simply calling `coerce` on a key doesn't make it show up in the help output -- this feels like a bit of an edge-case though, I think I'm probably just missing your use-case though.

**Comment by ljharb** [comment]
Conceptually, "foo.bar" is a single option. That I've configured the parser options - a secret implementation detail that I'm free to change at any time - to enable dot-notation parsing is irrelevant. I'm totally comfortable with writing my yargs code to validate/coerce "foo" all at once! However, "foo.bar" remains the only thing that should show up.

**Comment by bcoe** [comment]
@ljharb I think you should be able to get this behavior by disabling `dot-notation` entirely in the parser:

https://github.com/yargs/yargs-parser#dot-notation

You can do this by adding:

```
{
  "yargs": {
     "dot-notation": false
   }
}
```

to your application's `package.json`; this should allow you to then define "foo.bar" and describe it separately from "foo.snuh", etc. -- I think for your use-case it would be better not to rely on the _clever_ parsing of objects.

**Comment by ljharb** [comment]
That certainly seems like it would work; but then it would mean that users couldn't define nested objects in their config :-/
